### ISSUE TITLE
DNS stops working shortly after compose containers come up

### ISSUE THREAD

**Issue by dcastor** [docker/compose#6321]
## Description of the issue

This is my first inquiry/issue, so if I did something wrong in the formatting I apologize. Just let me know and I'll resolve it. 

I'm noticing a seemingly peculiar issue when using docker-compose for our containers on Centos 7.4.1708. Shortly after the container(s) come up, the DNS works for roughly a minute; after about a minute or so, DNS is no longer able to resolve any domains at all. Further, on Ubuntu 14.04, the containers seem to maintain DNS just fine. I am assuming it's related to some config on Centos, but I'm not too sure yet. 

At first, I noticed this spinning up a number of images in a single compose file, but after removing all contents, I noticed it happens with as little as one bare bones container, so it doesn't seem to be related to the combination of the containers.

I apologize for stripping some context out of the repo, but it's primarily internal. We noticed this error occurs with any image, though. 

## Context information (for bug reports)

**Output of `docker-compose version`**
```
docker-compose version
docker-compose version 1.22.0, build f46880fe
docker-py version: 3.4.1
CPython version: 3.6.6
OpenSSL version: OpenSSL 1.1.0f  25 May 2017
[sdb@sdbl10say1-sdb1-1-prd:~/archival]$ 

```

**Output of `docker version`**
```
Client:
 Version:      18.03.1-ce
 API version:  1.37
 Go version:   go1.9.5
 Git commit:   9ee9f40
 Built:        Thu Apr 26 07:20:16 2018
 OS/Arch:      linux/amd64
 Experimental: false
 Orchestrator: swarm

Server:
 Engine:
  Version:      18.03.1-ce
  API version:  1.37 (minimum version 1.12)
  Go version:   go1.9.5
  Git commit:   9ee9f40
  Built:        Thu Apr 26 07:23:58 2018
  OS/Arch:      linux/amd64
  Experimental: false

```

**Output of `docker-compose config`**
(Make sure to add the relevant `-f` and other flags)
```
  zookeeper:
    container_name: zookeeper
    expose:
    - '2888'
    - '3888'
    - '22'
    - '2181'
    image: [internal_repo]/zookeeper:latest
    logging:
      driver: none
version: '2.0'
```
/etc/resolv.conf contents of container: 
```
bash-4.2$ cat /etc/resolv.conf 
nameserver 127.0.0.11
options rotate timeout:2 attempts:2 ndots:2
```

## Steps to reproduce the issue

1. docker-compose up -f [compose_file]
2. Immediately after the lone container is up, do `docker-compose exec zookeeper /bin/bash`
3. ping `google.com` (for us, it's primarily internal DNS, but I noticed it happens for external as well)
4. Wait about 1-2 minutes
5. Retry. Eventually, it will fail. 

### Observed result
bash-4.2$ ping google.com
ping: google.com: Name or service not known
### Expected result
Successful DNS connection via embedded docker DNS routing. 

## Additional information

Problem OS: Centos 7.4.1708

Working OS: Distributor ID:	Ubuntu
Description:	Ubuntu 14.04.5 LTS
Release:	14.04
Codename:	trusty

**Comment by shin-** [comment]
Hi @dcastor 

`docker-compose` isn't directly in charge of container networking - that is still managed by the Docker Engine. You may want to create an issue on the [moby](https://github.com/moby/moby) repo instead.

cc @thaJeztah is this something you've seen before?

**Comment by thaJeztah** [comment]
hm, no doesn't look familiar at a glance. Possible suspect could be networkmanager.

If you are able to put the dockerd daemon in debug mode, and collect logs for situations this happens, possibly that would give more info

**Comment by dcastor** [comment]
Thanks for those tips. Sure, I can totally open this up elsewhere if needed. I am noticing a slew of DNS failed lookups after enabling debugging -- e.g.:

```
Nov  5 22:10:17 [internal].net systemd[1]: Starting Serial Getty on ttyS1...
Nov  5 22:10:18[internal].net dockerd[9069]: time="2018-11-05T22:10:18.881081455Z" level=debug msg="[resolver] read from DNS server failed, read udp 172.18.0.4:43742->10.253.128.226:53: i/o timeout"
```
That gives me some better info to search around for online to potentially find what setting is off here.

**Comment by dcastor** [comment]
Hi @thaJeztah, it does look like the majority of threads I find have you on them! I took a deeper look after spinning up a cluster and reproducing the issue. 

So I am noticing that I actually have a suffix that's being searched first. After that fails, it kind of strips this and then searches with an additional period. This also appears to be happening even for our internal domains that we search for. Not sure how literal that additional period is, but here is an example with a more friendly Yahoo domain search:

```Nov  7 00:02:04 local10-stt1-1-prd.local.internal.net dockerd[51586]: time="2018-11-07T00:02:04.716117607Z" level=debug msg="Name To resolve: yahoo.com.local.internal.net."
Nov  7 00:02:04 local10-stt1-1-prd.local.internal.net dockerd[51586]: time="2018-11-07T00:02:04.716176216Z" level=debug msg="[resolver] query yahoo.com.local.internal.net. (A) from 172.18.0.6:37563, forwarding to udp:10.253.128.235"
Nov  7 00:02:04 local10-stt1-1-prd.local.internal.net dockerd[51586]: time="2018-11-07T00:02:04.920108315Z" level=debug msg="[resolver] read from DNS server failed, read udp 172.18.0.7:45569->10.253.128.235:53: i/o timeout"
```

Next up, it tries w/o the appended domain:

```Nov  7 00:02:06 local10-stt1-1-prd.local.internal.net dockerd[51586]: time="2018-11-07T00:02:06.717792095Z" level=debug msg="Name To resolve: yahoo.com."
Nov  7 00:02:06 local10-stt1-1-prd.local.internal.net dockerd[51586]: time="2018-11-07T00:02:06.717854633Z" level=debug msg="[resolver] query yahoo.com. (A) from 172.18.0.6:43859, forwarding to udp:10.253.128.235"
Nov  7 00:02:06 local10-stt1-1-prd.local.internal.net dockerd[51586]: time="2018-11-07T00:02:06.918508652Z" level=debug msg="[resolver] read from DNS server failed, read udp 172.18.0.7:44852->10.253.128.226:53: i/o timeout"
Nov  7 00:02:06 local10-stt1-1-prd.local.internal.net dockerd[51586]: time="2018-11-07T00:02:06.918585807Z" level=debug msg="[resolver] query stt1.local.internal.net. (A) from 172.18.0.7:37582, forwarding to udp:10.253.128.227"
Nov  7 00:02:06 local10-stt1-1-prd.local.internal.net dockerd[51586]: time="2018-11-07T00:02:06.918508990Z" level=debug msg="[resolver] read from DNS server failed, read udp 172.18.0.7:51054->10.253.128.226:53: i/o timeout"
```

I'll keep poking around.

**Comment by dcastor** [comment]
So we took a deeper look at this one with tcpdump. We can confirm that the traffic is indeed going out, but on the Ubuntu box, the outbound address is essentially the hostname, so it knows how to route the traffic back in; thus we get successful pings. On Centos, however, the source address that sends out is the bridge network IP, which doesn't route back to the box, so we can see the traffic go out but it has no idea how to route it back in. 

One other peculiar thing was that there are a number of docker iptable rules on the Ubuntu box, whereas there are zero on the centos box for any docker user. Not sure why that is. 

Will post back with an update if we find the source. Thanks!

**Comment by dcastor** [comment]
Follow up:

End result was that somehow, somewhere `--iptables` was disabled. As per documentation, it is supposed to be true, but it clearly wasn't inserting the rules. I ended up setting this both in /etc/docker/daemon.json and via DOCKER_OPTS (assuming former was the fix), restarting daemon, and then starting a container. This revealed that the outbound source in tcpdump went from the container IP (unable to route back in to the container) to the host via the bridge network. This was able to route back in, but then I had to set `iptables -P FORWARD DROP` to `iptables -P FORWARD ACCEPT`. Oddly, the first time I did this crashed the docker daemon. I repeated above steps and this time it worked. Ping success within container!

Closing this.
